### DISCUSSION TITLE
Change Canvas Origin (Center Origin)

### DISCUSSION THREAD

**Discussion by 201flaviosilva** [fabricjs/fabric.js#7778] [Q&A] [answered]
Hello everyone, once again I'm here to ask weird questions 😅

This time my question is, is there any way in fabricjs to change the origin point of the screen to the center of the screen, for example.

The screen is 800x600, its origin point would be 400x300.
And the objects would be designed taking into account the new orientation and following the logic of traditional Cartesian axes (something like Unity does).

Demo image:

<img width="335" alt="Captura de ecrã 2022-03-08, às 10 28 01" src="https://user-images.githubusercontent.com/55815265/157220222-1d2f6c3b-ef57-4569-bd45-aaf9e81d9bc5.png">

I did the same example of changing the origin, both in canvas and in fabricjs, it is worth mentioning that in fabricjs to drag an object it is necessary to go to the "original" point of the object, and also objects with negative axes are not being displayed (I would say which is because fabricjs assumes they are negative and therefore, to save performance, doesn't draw them).

Here are the examples:

- [Canvas](https://jsfiddle.net/201flaviosilva/bpyL670j/): https://jsfiddle.net/201flaviosilva/bpyL670j/
- [Fabric](https://jsfiddle.net/201flaviosilva/21ou06xv/1/): https://jsfiddle.net/201flaviosilva/21ou06xv/1/

---
(Sorry for my bad English, and also if I was confused, and thanks to everyone who answers (if anyone answers 🥲)) :)

**Comment by ShaMan123** [comment, marked as answer]
fabric is designed in such a way that canvas (0,0) is the top left corner.
Objects' (0,0) however are their parent's origin point. If the parent is an object then it's the parent's center point, if it's canvas then top left point of canvas.
I agree it's confusing and bad design.

This is what you are looking for I believe
```diff
- canvas.getContext().transform(1, 0, 0, -1, halfCWidth, halfCHeight);
+ canvas.setViewportTransform([1, 0, 0, -1, halfCWidth, halfCHeight]);
 // you are right that fabric checks if objects are off screen and skips them if not, it is still buggy
// this overrides the check
+ canvas.skipOffscreen = false; 
```

https://codepen.io/shaman123/pen/RwjmKNB

**Reply by ShaMan123** [reply]
You can bridge the difference with a method

**Reply by ShaMan123** [reply]
Or better
```diff
- canvas.getContext().transform(1, 0, 0, -1, halfCWidth, halfCHeight);
+ canvas.setViewportTransform([1, 0, 0, -1, halfCWidth, halfCHeight]);
+ canvas.skipOffscreen = false;  // you are right that fabric checks if objects are off screen and skips them if not, it is still buggy
```

https://codepen.io/shaman123/pen/RwjmKNB

**Reply by 201flaviosilva** [reply]
Thank you very much, it was exactly what I was looking for.

To be honest, it doesn't confuse me at all that 0x0 is in the top left corner, I'm just trying to do this for a specific project :)

**Reply by ShaMan123** [reply]
There's a PR that should fix some cases of the viewport transform bug #7515

**Reply by 201flaviosilva** [reply]
I mean it's still not quite perfect, when an entire object this happens 😅:

<img width="164" alt="Captura de ecrã 2022-03-08, às 11 09 06" src="https://user-images.githubusercontent.com/55815265/157227054-c667b240-e64b-4b15-9400-5b56b9840b88.png">

**Reply by ShaMan123** [reply]
Once it's fixed you could drop the offscreen flag 
Try patching fabric to see if it fixes your use case

**Reply by 201flaviosilva** [reply]
Cool, I'll keep an eye out, and once again very grateful :)

**Reply by ShaMan123** [reply]
It's because of the negative scaling. Submit a issue after trying the patch.
Look into `calcOCoords`, `drawBroders`, `drawControls`

**Reply by 201flaviosilva** [reply]
Sorry, I don't think I'm keeping up, what do you mean? 😅

**Reply by ShaMan123** [reply]
Negative scaling in viewportTransform is poorly supported.
If you want to try to fix it by submitting a PR look into the following methods `calcOCoords, drawBroders, drawControls`.
That's what needs to be fixed.

**Reply by 201flaviosilva** [reply]
Ahhh OK, no, I think I'll just create an issue and when the devs have time they will fix it :)

I'm still too noob to jump into a framework of this size :)

**Reply by 201flaviosilva** [reply]
Issue created: https://github.com/fabricjs/fabric.js/issues/7779 😅

**Reply by ShaMan123** [reply]
I disagree with your approach.
Contributing is what drives open source.

**Comment by chinyifei** [comment]
Fabric cannot change the Canvas Origin, you may need to change the coordinate system using
